src/os/bluestore/BlueStore.cc

```diff
@@ -14254,14 +14254,36 @@ void BlueStore::_do_write_big(
     }
     bufferlist t;
     blp.copy(l, t);
-    wctx->write(offset, b, l, b_off, t, b_off, l, false, new_blob);
-    dout(20) << __func__ << " schedule write big: 0x"
-      << std::hex << offset << "~" << l << std::dec
-      << (new_blob ? " new " : " reuse ")
-      << *b << dendl;
+
+    // Zero detection -- fine grain
+    //  scan through t, the original bufferlist, and look at smaller allocation units (alloc_unit)
+    //  check from byte 0 -> alloc_unit
+    //  min_alloc_size
+
+    // Fine-grain detection only occurs if bufferlist t contains non-zero values.
+    if (!t.is_zero()) {
+      bufferlist non_zeroes; // will contain only relevant parts of t
+      bufferlist::iterator iter = t.begin();
+      while (iter != t.end()) {
+        bufferlist alloc_unit; // will contain a section of t, size min_alloc_size
+        iter.copy(min_alloc_size, alloc_unit); // copy advances the iterator fowared by min_alloc_size
+        if (!alloc_unit.is_zero()) {
+          // only keep the data that is non_zero
+          non_zeroes.append(alloc_unit);
+        }
+      }
+      // non_zeroes contains only the relevant metadata from t. This is what's written to the disk.
+      wctx->write(offset, b, l, b_off, non_zeroes, b_off, l, false, new_blob);
```

Comment:
I'm pretty skeptical about this implementation's correctness. Let's run through the following example:
1) original bufferlist t is as follows: t = <0: 4K, non-zero>, <1: 4K, zero>, <2: 4K, non-zero>
2) After zero detection loop one gets: non_zero = <0: 4K, non-zero>, <2: 4K, non-zero>
3) Finally single wctx->write (with all the unmodified lengths and offsets)  is invoked - how would it properly dissect the incoming bufferlist and produce the required gap between units 0 and 2? Highly likely it would just write units 0 and 2 continuously hence breaking the user data. The follow block sequence is presumably a result: <0: 4K, non-zero>, <2: 4K, 
non-zero>, <1: 4K, zero>

Reply:
Ah, I see what you mean, I think this is one of those cases where I've been staring at the implementation for so long, that I didn't consider this case. What I intended was for the space to be logically extended between <0 : 4k, non_zero> and <2: 4K, non-zero>, but I see now that they are getting written consecutively.

I will work on fixing this today.